### PR TITLE
Add a blocklist of sites we don't launch on

### PR DESCRIPTION
**Pull request by seanh**
Add a blocklist of sites we don't launch on
The blocklist is configured in production.ini like so:

    h.blocklist = {
        "seanh.cc": {},
        "finance.yahoo.com": {},
        "twitter.com": {},
        "*.google.com"
    }

The Chrome extension handles the blocklist in the same way that it
handles other launch errors (e.g. when the user hasn't given the
extension permission to annotate file:/// URLs): it adds a (!) error
badge to the "browser action" button, if the button is clicked again it
opens a "help" page saying that the site isn't supported yet.

It doesn't seem to be possible to pass data into the Chrome extension's
help pages (as far as I can tell) so the error is just "Hypothesis
doesn't work on this page yet" and not e.g. "Hypothesis doesn't work on
twitter.com".

The bookmarklet handles the blocklist in the same way that it handles
other launch errors (e.g. trying to annotate a PDF when PDF.js isn't
available): it uses window.alert() to show an error.

### CODE REVIEW

**Comment by seanh** [comment]
@judell wanted to make the blocklist editable by users without doing a deploy. That's not going to be included in this initial PR, but if we did do it I'd say store the blocklist in the db instead of the config file and add a simple API for viewing the list of blocked sites, blocking a site, and unblocking a site.

**Comment by seanh** [comment]
What do people think about tests for this? I can certainly write tests for the blocklist config setting validation in Python. But tests for `embed.js`?

I think the two things remaining to be done here are tests and replacing `window.alert()`

**Comment by seanh** [comment]
> Biggest problem is using window.alert() is ugly. But how to make this better? Since we're deliberately
not injecting our Angular app. Possible solution: inject a different Angular app instead! A really simple standalone one that just requires toastr and displays one message.

Err ... Or just call toastr direct from js without angular (facepalm smiley)

**Comment by judell** [comment]
Wildcards?

We will almost certainly have to say *.google.com for example.

**Comment by seanh** [comment]
Wildcards can be done fairly easy. I'm having more trouble getting a nice way of showing the notification working

**Comment by seanh** [comment]
The toast notification in commit <https://github.com/hypothesis/h/commit/664d70887d640b07a51c4b05e9f1f4abc71144a2> is working fine with the bookmarklet, but doesn't work with the Chrome extension on any site (https and csp or not): Uncaught ReferenceError: toastr is not defined at embed.js line 49.

There's no other error (e.g from the browser refusing to inject the script). If I put a breakpoint right before the line where it crashes and inspect the HTML, the `<script>` for toastr _has_ been inserted in the `<head>` (though I guess that doesn't necessarily mean the script has been executed).

I tried vendoring the files as well. So now when I inspect the HTML in the debugger I can see a `<script>` for `chrome-extension://oldbkmekfdjiffgkconlamcngmkioffd/public/scripts/vendor/toastr.min.js` has been injected, and if I go to that URL I do see the js file. Same error though: toastr is undefined.

**Comment by csillag** [comment]
> Uncaught ReferenceError: toastr is not defined at embed.js line 49.
> [...] 
>  the `<script>` for toastr _has_ been inserted in the `<head>`

I haven't really tried to debug this, just a question: you are surely aware that there are two different spaces here? Part of the code of the chrome extension runs in a different scope, separate from the HTML page. That could explain not finding the libraries you wanted to load. (But of course it's also possible that all the code you are talking about is already running inside the page; as I said, I haven't looked too closely.)

**Comment by seanh** [comment]
@csillag Thanks, I think you're onto it. Not tested a solution yet but I think the problem is probably that when injected by the chrome extension (sidebar-inject.jr does it with chrome.tabs.executeScript()) embed.js runs as a _content script_ and apparently content scripts cannot "Use variables or functions defined by web pages or by other content scripts". <https://developer.chrome.com/extensions/content_scripts>

So that would explain it, and why the same code works fine when injected by the bookmarklet.

Basically I don't know much about Chrome extensions (never done one before), thinking this was a problem that should be solvable quickly I just kept going at it for half a day or more, and never bothered to step back and go away and learn about Chrome extensions then come back. Rookie mistake.

I'll come back to this when I've figured a workaround, then we can test whether this was in fact the problem

**Comment by tilgovi** [comment]
The content script can't really "inject" code, it can add a script tag. That script tag is run in the same context as any others, not the content script context.

**Comment by seanh** [comment]
@tilgovi Yep :) Thanks. I'm also rewriting this so that the Chrome extension will do what it currently does for other kinds of unsupported pages (e.g. when it doesn't have permission to annotate a `file:///` URL): shows an errored badge on the button, if you click it again opens a help page in a new tab. And the bookmarklet will do what _it_ currently does for other errors, which is use the browser's `alert()` function.

**Comment by seanh** [comment]
Just force pushed this with a completely new version, and edited the first comment at the top of this page.

**Comment by seanh** [comment]
untitled screencast - cropped

**Comment by seanh** [comment]
What does this still need?

- [ ] Feature flag? Seems unnecessary to me, if you've no blocklist setting in your ini nothing will be blocked
- [ ] Tests
- [ ] Wildcard support (regexes is probably easiest)

**Comment by seanh** [comment]
Think this can be reviewed now

Thoughts:

- Is the button badge then help page the right thing to do here? We may be able to do something nicer (Chrome notifications API maybe? Or a popup window on the button?) but maybe that belongs in a separate PR that changes this behaviour for the other errors we already handle as well

- No tests for the bookmarklet

- Had to add `h.parseUrl` to make the function testable. Also there may be a better place for this function.

**Comment by seanh** [comment]
Oh, still needs wildcard support

File: h/browser/chrome/lib/sidebar-injector.js [hunk 1]
```diff
@@ -1,6 +1,40 @@
...
   'use strict';
 
+  /* Parse the given URL and return an object with its different components.
+   *
+   * Any or all of the components returned may be undefined.
+   * For example for the URL "http://twitter.com" port, path query and anchor
+   * will be undefined.
+   *
+   */
+  function parseUrl(url) {
```
Comment (tilgovi) [inline h/browser/chrome/lib/sidebar-injector.js]: https://developer.mozilla.org/en-US/docs/Web/API/URL

Comment (seanh) [inline h/browser/chrome/lib/sidebar-injector.js]: Yeah, I did use URL originally but it didn't seem to work in PhantomJS

**Comment by seanh** [comment]
The search for current URL in blocklist code is duplicated in two places currently, would be nice if we could move it into somewhere shared. One of the two places where it's duplicated is `embed.js` though

**Comment by seanh** [comment]
Oops, got a `.only(` in there

**Comment by seanh** [comment]
The biggest remaining problem I have with this pr is this function, in the Chrome extension:

```javascript
    /* Return true if the given url is blocked by the given blocklist. */
    function isBlocked(url, blocklist) {
      var regexSpecialChars = '^$.+?=|\/()[]{}'; //  '*' deliberately omitted.
      for (var pattern in blocklist) {
        if (blocklist.hasOwnProperty(pattern)) {
          // Escape regular expression special characters.
          for (var i = 0; i < regexSpecialChars.length; i++) {
            var c = regexSpecialChars.charAt(i);
            pattern = pattern.replace(c, '\\' + c);
          }

          // Turn * into .* to enable simple patterns like "*.google.com".
          pattern = pattern.replace('*', '.*');

          // Blocklist patterns should match from the start of the URL.
          // This means that "google.com" will _not_ match "mail.google.com",
          // for example. (But "*.google.com" will match it.)
          pattern = '^' + pattern;

          if (url.match(pattern)) {
            return true;
          }
        }
      }
      return false;
    }
```

That's fairly complicated and it's duplicated (without wildcard support yet) in `embed.js` for the bookmarklet. How can I share this function between the Chrome extension and the bookmarklet?

**Comment by seanh** [comment]
@tilgovi @nickstenning The last commit introduces some test fails that I don't understand. Any clues?

The commit gets rid of some warnings printed in the test output. Each time it makes an XMLHttpRequest for blocklist.json that fails it prints out a warning. It's actually fine - the code can handle this and the tests pass. But messes up the test runner output. So I just made blocklist.json available to all the tests. But for some reason this breaks a few.

**Comment by seanh** [comment]
Tests should be passing now, but still need to get the wildcard support in the bookmarklet..

**Comment by seanh** [comment]
That's the wildcard support done and duplication removed, and rebased. Ready for review!

**Comment by tilgovi** [comment]
Got a lint failure here.

**Comment by tilgovi** [comment]
ftfy
